### PR TITLE
fix(controller): JSON-unmarshal marshaled expression template before evaluating

### PR DESCRIPTION
**Pull request by mac9416**
Signed-off-by: Michael Crenshaw <michael@crenshaw.dev>

Fixes https://github.com/argoproj/argo-workflows/issues/6260 and https://github.com/argoproj/argo-workflows/issues/6325

### The problem

FastTemplate plucks the `{{=whatever}}` expression template from the stringified JSON representation of the (Argo script/container) template and replaces it with whatever the expression evaluates to.

So for example, you might start with:

```json
"env": [
  {
    "name": "A",
    "value": "{{=toJson('a')}}"
  }
]
```

That becomes

```json
"env": [
  {
    "name": "A",
    "value": ""a""
  }
]
```

When it should be

```json
"env": [
  {
    "name": "A",
    "value": "\"a\""
  }
]
```

### Testing

I tested with this workflow, which would have previously failed:

```yaml
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  name: expression-template-double-quotes
spec:
  entrypoint: main
  templates:
    - name: main
      script:
        env:
          - name: A
            value: '{{="test"}}'
        image: debian:9.4
        command: [bash]
        source: |
          echo "$A"
```

And this workflow, which would have previously failed:

```yaml
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  name: expression-template-double-quotes
spec:
  entrypoint: main
  templates:
    - name: main
      script:
        env:
          - name: A
            value: '{{=toJson('a')}}'
        image: debian:9.4
        command: [bash]
        source: |
          echo "$A"
```

### CODE REVIEW

File: docs/running-locally.md [hunk 1]
```diff
@@ -5,32 +5,43 @@
 * [Go](https://golang.org/dl/) (The project currently uses version 1.15)
 * [Yarn](https://classic.yarnpkg.com/en/docs/install/#mac-stable)
 * [Docker](https://docs.docker.com/get-docker/)
-* [Kustomize](https://github.com/kubernetes-sigs/kustomize/blob/master/docs/INSTALL.md)
+* [Kustomize](https://kubectl.docs.kubernetes.io/installation/kustomize/)
```
Comment (mac9416) [inline docs/running-locally.md:8]: Their docs moved.

File: hack/recurl.sh [hunk 2]
```diff
@@ -4,10 +4,12 @@ set -eux -o pipefail
 file=$1
 url=$2
 
-# loop forever
-while ! curl -L -o "$file" -- "$url" ;do
-  echo "sleeping before trying again"
-  sleep 10s
-done
+if [ ! -f "$file" ]; then
```
Comment (mac9416) [inline hack/recurl.sh:7]: I had an issue where this fell into an endless loop of failing to write a file that was already there.

File: util/template/template_test.go [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+	t.Run("Expression", func(t *testing.T) {
+		t.Run("Valid", func(t *testing.T) {
+			template, err := NewTemplate("{{=\\\"test\\\"}}")
+			assert.NoError(t, err)
+			r, err := template.Replace(map[string]string{}, true)
+			assert.NoError(t, err)
+			assert.Equal(t, "test", r)
+		})
+	})
+}
```
Comment (simster7) [inline util/template/template_test.go:19]: Can we get a control test? One for `value: "{{='test'}}"  # evaluates to `test``?

**Review by simster7 — CHANGES_REQUESTED** [review]
If you want to separate your changes to `docs/running-locally.md` and `hack/recurl.sh` from the rest into a separate PR, I can merge those immediately.

**Comment by mac9416** [comment]
@simster7 I haven't really thought through the behavior of expr w.r.t. escaped backslashes... I might add another replace line for `\\` -> `\`.

**Comment by mac9416** [comment]
@simster7 I changed from a naive string replace solution to what I think is a more robust JSON-unmarshal approach.

Apologies if there are test failures, I'm on a machine with an incomplete dev env right now.

I'll open a PR with just the docs/hack changes.

**Comment by mac9416** [comment]
Note to self: add a test for templates that contain newlines. Seems reasonable to break a long expression template across a couple lines, and I think the JSON-encoding would break the expression evaluation.

**Comment by mac9416** [comment]
I don't know how to interpret this error when running `make pre-commit -B` locally:

```
?       github.com/argoproj/argo-workflows/v3/workflow/verify   [no test files]
FAIL
make: *** [test] Error 1
```

Must be a local config issue, because I don't see it in the CI tests.

**Comment by mac9416** [comment]
Just realized the output needs to be marshaled as a JSON string to avoid errors when unmarshaling the template again.

The lack of that step to marshal as a string causes this to fail in 3.1.1:

```
{{=toJson(map(0..4, {sprig.randNumeric(5)}))}}
```

**Comment by mac9416** [comment]
Methinks the tests are flaky.

File: util/template/expression_template.go [hunk 4]
```diff
@@ -17,12 +18,22 @@ func init() {
 }
 
 func expressionReplace(w io.Writer, expression string, env map[string]interface{}, allowUnresolved bool) (int, error) {
-	if _, ok := env["retries"]; !ok && hasRetries(expression) && allowUnresolved {
+	// The template is JSON-marshaled. This JSON-unmarshals the expression to undo any character escapes.
```
Comment (alexec) [inline util/template/expression_template.go:21]: we need to be very careful with expressions, I've seen many cases of patching specific issues, that just produce new issues

Comment (mac9416) [reply util/template/expression_template.go:21]: Yep, there's a lot going on here.

I would _like_ to be more expressive with the change. Basically, it's "grab part of a JSON string, unescape it, evaluate it, and escape the result before re-inserting it in the string".

But the code is much more difficult to read.

Comment (mac9416) [reply util/template/expression_template.go:21]: btw, it looks like example workflows can be invoked as e2e tests... if I can figure out that system, I'll add some example workflows which test at least these bugs.

**Comment by alexec** [comment]
I'm not sure what the correct solution is here. Basically (if I remember) the templating works by converting the workflow into JSON text, running the substitution, and then converting it back. It would be leaky to pretend that is not the case. So I wonder if we should introduced some escaping function?

```
{{=escape(toJSON('a'))}}
```

**Comment by mac9416** [comment]
I'm unsure what you mean by "leaky".

I can't imagine a case where we could _not_ want to JSON-stringify the expression result, because that result will _always_ be inerpolated into a JSON string.

The only thing we need to do besides stringify the result is to strip the leading and trailing `"`. They're implicit, provided by the string which previously contained the expression.

**Comment by alexec** [comment]
What I mean by "leaky" is that "all abstracting are leaky". Our templating system is a good example, there are edge-case failures which can only be diagnosed by understanding the underlying abstraction and that it does not work at a field level (as I think many would I assume) but by marshalling and unmarshalling the template to/from text/JSON.

This issue is a good example of leakage caused by a mismatch between what the abstraction makes you expect, and what it actually is.

**Comment by mac9416** [comment]
Yep I follow that. I'm still not sure I can see the potential leak caused by this unmarshal-execute-marshal approach. As long as `expressionReplace` is only ever called with an `expression` which is part of a string in a stringified JSON template, I don't think there is a possible leak.

**Comment by alexec** [comment]
@simster7 opinion?

**Comment by simster7** [comment]
I mostly agree with @alexec's reasoning on "leakiness" overall, but if you're able to write exhaustive tests on your approach I think we should be fine

**Comment by mac9416** [comment]
@simster7 cool I'ma try to figure out the e2e testing setup tonight and add some.

Exhaustive might be a bit difficult with a tool this powerful. XD Are you thinking just a _lot_ of tests, covering a really wide variety of expressions? Or is there something more specific you have in mind?

**Comment by simster7** [comment]
I think unit tests are fine... I don't see a need for E2E tests in this case. I think just a handful of tests covering every edge case you can think of should be enough

**Comment by mac9416** [comment]
This isn't really a bug, because it requires the developer to behave ridiculously, but it does give me a chuckle:

```yaml
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  generateName: the-worst-way-to-template-
spec:
  entrypoint: main
  templates:
    - name: main
      script:
        image: debian:9.4
        command: [bash]
        env:
          - name: A
            value: "{{='1' + sprig.substr(0, 1, toJson('')) + '},' + sprig.substr(0, 22, toJson({name: 'B', value: '2'})) + '"
          - name: C
            value: "'}}"
        source: |
          echo "$A$B$C"
```

Output: `12`

Anyhow, I'ma work on some more tests over the next few days and then I'll mark this as ready for review. :-)

**Comment by mac9416** [comment]
I haven't forgotten this.

The last thing I need to do is analyze each call of `template.Replace` to make sure that what's being passed in is valid JSON (since the expression template code will assume it is receiving JSON). At the moment, this assumption seems to be true of every call except for one. For that one, I will change the code to serialize the input string into a JSON string for templating and then deserialize it for use as a string.

I wish I could change the signature of `template.Replace` to accept `interface{}` and JSON serialize it internally. But to do that, I would need a much deeper understanding of error handling in the context of each place that currently does that JSON serialization before running the template engine.

This might seem like overkill. But as expression templates are currently implemented, they're completely broken for important use cases (generating JSON containing string keys and values) and permit potentially dangerous handling of input in ways the workflow author might not anticipate.

**Comment by alexec** [comment]
I wonder if we're doing this wrong. Instead of converting to JSON and then templating the lot, what about if we convert to `map[string]interface{}` and then iterate over the value templating them one-by-one. This could fix it.

**Comment by mac9416** [comment]
Been afk a few days, haven’t had time to work on this.

But yeah, I think field-wise templating would be best. Not sure if there would be any performance impacts? I’ll look at what I think that would take when I come back to this (hopefully Thursday evening).

I think this PR is _technically_ in a mergeable state, but I wanted to run some tests templating in the metrics fields since I have changes in that area.

**Comment by alexec** [comment]
@markterm what help do you need from us to get this over the line?

**Comment by mac9416** [comment]
@alexec I think you might have auto-completed the wrong name, but happy for more help if @markterm has time.  ;-)

The only thing I want to do is to run a workflow or two with templating in the metrics fields (that's where I made the most invasive changes).

Unfortunately, I can't bill this work to my employer (big yuck), so I'm having to push it forward on weekends. If an Argo dev could verify that metrics templating works fine, I'll be happy with this PR. Otherwise, I'll do my best to carve out a few minutes some evening to give it a test.

**Comment by mac9416** [comment]
I've been reading the unit test output incorrectly. I see there's a metrics-related failure. Maybe strconv.Quote and JSON marshaling have a bad interaction. Will look at it.

**Comment by mac9416** [comment]
@alexec now that I've figured out how to read the test logs, I've knocked out a couple of other bugs.

This one has me stumped:

```
--- FAIL: TestGetPodByNode (0.10s)
    operator_test.go:3534: 
        	Error Trace:	operator_test.go:3534
        	Error:      	Expected value not to be nil.
        	Test:       	TestGetPodByNode
```

**Comment by mac9416** [comment]
It passes locally in GoLand, so... flakiness?

**Comment by alexec** [comment]
That test is flakey. See #6458

**Comment by mac9416** [comment]
Sweet. Looks like I missed a codegen thingy again. Will run it tomorrow.

**Comment by mac9416** [comment]
@alexec looks like this is gonna pass all the checks. I still don't love the code, but it'll fix the issue until we can go to your per-field evaluation approach.

File: examples/expression-reusing-verbose-snippets.yaml [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+# be used elsewhere.
+apiVersion: argoproj.io/v1alpha1
+kind: Workflow
+metadata:
+  generateName: expression-reusing-verbose-snippets-
+spec:
+  arguments:
+    parameters:
+      - name: weather
+        # The base64 string is th JSON: {"temps": [34, 27, 15, 57, 46]}
```
Comment (mac9416) [inline examples/expression-reusing-verbose-snippets.yaml:12]:
```suggestion
        # The base64 string is this JSON: {"temps": [34, 27, 15, 57, 46]}
```

**Review by alexec — APPROVED** [review]

**Comment by SVilgelm** [comment]
Could you please bump new version with this fix?
